File: 30_mini_projects/17_polyglot_forge/src/loop.py
# src/loop.py
# Event loop minimal : macrotask queue + microtask queue.
# Lit tests/scenario.json (chemin passe en argv[1]), imprime une trace ligne par ligne.

import json, sys
from collections import deque

path = sys.argv[1] if len(sys.argv) > 1 else None
if not path:
    print("usage: python3 loop.py <scenario.json>", file=sys.stderr); sys.exit(2)
with open(path, encoding="utf-8") as f:
    scenario = json.load(f)

macros = deque()
micros = deque()
out = []

def schedule(evt):
    if evt["kind"] == "macro":
        macros.append(evt)
    elif evt["kind"] == "micro":
        micros.append(evt)

for evt in scenario["initial"]:
    schedule(evt)

tick = 0
while macros or micros:
    while micros:
        m = micros.popleft()
        out.append(f"t={tick} micro {m['name']}")
        for c in m.get("enqueue", []):
            schedule(c)
    if macros:
        m = macros.popleft()
        out.append(f"t={tick} macro {m['name']}")
        for c in m.get("enqueue", []):
            schedule(c)
        tick += 1

sys.stdout.write("\n".join(out) + "\n")

File: 30_mini_projects/17_polyglot_forge/src/loop.js
// src/loop.js
// Event loop minimal : macrotask queue + microtask queue.
// Lit tests/scenario.json (chemin passe en argv[2]), imprime une trace ligne par ligne.

const fs = require('fs')
const path = process.argv[2]
if (!path) { console.error('usage: node loop.js <scenario.json>'); process.exit(2) }
const scenario = JSON.parse(fs.readFileSync(path, 'utf8'))

const macros = []
const micros = []
const out = []

function schedule(evt) {
  if (evt.kind === 'macro') macros.push(evt)
  else if (evt.kind === 'micro') micros.push(evt)
}

for (const evt of scenario.initial) schedule(evt)

let tick = 0
while (macros.length || micros.length) {
  while (micros.length) {
    const m = micros.shift()
    out.push(`t=${tick} micro ${m.name}`)
    for (const c of (m.enqueue || [])) schedule(c)
  }
  if (macros.length) {
    const m = macros.shift()
    out.push(`t=${tick} macro ${m.name}`)
    for (const c of (m.enqueue || [])) schedule(c)
    tick += 1
  }
}

process.stdout.write(out.join('\n') + '\n')
